packages/test/src/test-integration-workflows-with-recorded-logs.ts

```diff
@@ -0,0 +1,298 @@
+              instanceOf: workflow.WorkflowNotFoundError,
+            })) as workflow.WorkflowNotFoundError;
+            this.t.is(err.message, 'workflow execution already completed');
+          } else {
+            await executeUpdate;
+          }
+          break;
+        }
+      }
+      const handlerFinished = await handle.result();
+      const unfinishedHandlerWarningEmitted =
+        recordedLogs[handle.workflowId] &&
+        recordedLogs[handle.workflowId].findIndex((e) => this.isUnfinishedHandlerWarning(e)) >= 0;
+      return [handlerFinished, unfinishedHandlerWarningEmitted];
+    });
+  }
+
+  isUnfinishedHandlerWarning(logEntry: LogEntry): boolean {
+    return (
+      logEntry.level === 'WARN' &&
+      new RegExp(`^Workflow finished while an? ${this.handlerType} handler was still running\\.`).test(logEntry.message)
+    );
+  }
+}
+
+export const unfinishedHandlersWithCancellationOrFailureUpdate = workflow.defineUpdate<void>(
+  'unfinishedHandlersWithCancellationOrFailureUpdate'
+);
+export const unfinishedHandlersWithCancellationOrFailureSignal = workflow.defineSignal(
+  'unfinishedHandlersWithCancellationOrFailureSignal'
+);
+
+export async function runUnfinishedHandlersWithCancellationOrFailureWorkflow(
+  workflowTerminationType: 'cancellation' | 'failure'
+): Promise<never> {
+  workflow.setHandler(unfinishedHandlersWithCancellationOrFailureUpdate, async () => {
+    await workflow.condition(() => false);
+    throw new Error('unreachable');
+  });
+
+  workflow.setHandler(unfinishedHandlersWithCancellationOrFailureSignal, async () => {
+    await workflow.condition(() => false);
+    throw new Error('unreachable');
+  });
+
+  if (workflowTerminationType === 'failure') {
+    throw new workflow.ApplicationFailure('Deliberately failing workflow with an unfinished handler');
+  }
+  await workflow.condition(() => false);
+  throw new Error('unreachable');
+}
+
+test('unfinished update handler with workflow cancellation', async (t) => {
+  await new UnfinishedHandlersWithCancellationOrFailureTest(t, 'update', 'cancellation').testWarningIsIssued();
+});
+
+test('unfinished signal handler with workflow cancellation', async (t) => {
+  await new UnfinishedHandlersWithCancellationOrFailureTest(t, 'signal', 'cancellation').testWarningIsIssued();
+});
+
+test('unfinished update handler with workflow failure', async (t) => {
+  await new UnfinishedHandlersWithCancellationOrFailureTest(t, 'update', 'failure').testWarningIsIssued();
+});
+
+test('unfinished signal handler with workflow failure', async (t) => {
+  await new UnfinishedHandlersWithCancellationOrFailureTest(t, 'signal', 'failure').testWarningIsIssued();
+});
+
+class UnfinishedHandlersWithCancellationOrFailureTest {
+  constructor(
+    private readonly t: ExecutionContext<Context>,
+    private readonly handlerType: 'update' | 'signal',
+    private readonly workflowTerminationType: 'cancellation' | 'failure'
```

Comment:
TODO (blocking PR): this should test continue-as-new

Reply:
I've added coverage for CAN now. The tests confirm that the warnings, and waiting on `allHandlersFinished` behave as expected in the various cases of handlers being interrupted by workflow return, CAN, failure, and cancellation.